src/machines/onboardingMachine.tsx

```diff
@@ -1,47 +1,99 @@
 
 const debug: Debugger = Debug(`${APP_ID}:onboarding`);
 
-const onboardingToSAgreementGuard = (ctx: OnboardingMachineContext, _event: OnboardingMachineEventTypes) =>
-  ctx.termsConditionsAccepted && ctx.privacyPolicyAccepted;
-
-const onboardingPersonalDataGuard = (ctx: OnboardingMachineContext, _event: OnboardingMachineEventTypes) => {
-  const {firstName, lastName, emailAddress} = ctx.personalData;
-  return firstName && firstName.length > 0 && lastName && lastName.length > 0 && emailAddress && EMAIL_ADDRESS_VALIDATION_REGEX.test(emailAddress);
-};
-
-const onboardingPinCodeSetGuard = (ctx: OnboardingMachineContext, _event: OnboardingMachineEventTypes) => {
-  const {pinCode} = ctx;
-  return pinCode && pinCode.length === 6;
-};
-
-const onboardingPinCodeVerifyGuard = (ctx: OnboardingMachineContext, event: NextEvent) => {
-  return onboardingPinCodeSetGuard(ctx, event) && ctx.pinCode === event.data;
+const states: OnboardingStatesConfig = {
+  showIntro: {
+    on: {
+      NEXT: OnboardingMachineStateType.showProgress,
+    },
+  },
+  showProgress: {
+    on: {
+      NEXT: [
+        {cond: ({currentStep}) => currentStep === 1, target: OnboardingMachineStateType.enterName},
+        {cond: ({currentStep}) => currentStep === 2, target: OnboardingMachineStateType.enterPinCode},
+      ],
+      PREVIOUS: [
+        {cond: ({currentStep}) => currentStep === 1, target: OnboardingMachineStateType.showIntro},
+        {
+          cond: ({currentStep}) => currentStep === 2,
+          target: OnboardingMachineStateType.enterCountry,
+          actions: assign({currentStep: 1}),
+        },
+      ],
+    },
+  },
+  enterName: {
+    on: {
+      NEXT: OnboardingMachineStateType.enterEmailAddress,
+      PREVIOUS: OnboardingMachineStateType.showProgress,
+      SET_NAME: {actions: assign({name: (_, event) => event.data})},
+    },
+  },
+  enterEmailAddress: {
+    on: {
+      NEXT: OnboardingMachineStateType.enterCountry,
+      PREVIOUS: OnboardingMachineStateType.enterName,
+      SET_EMAIL_ADDRESS: {actions: assign({emailAddress: (_, event) => event.data})},
+    },
+  },
+  enterCountry: {
+    on: {
+      NEXT: {
+        target: OnboardingMachineStateType.showProgress,
+        actions: assign({currentStep: 2}),
+      },
+      PREVIOUS: OnboardingMachineStateType.enterEmailAddress,
+      SET_COUNTRY: {actions: assign({country: (_, event) => event.data})},
+    },
+  },
+  enterPinCode: {
+    on: {
+      NEXT: OnboardingMachineStateType.verifyPinCode,
+      PREVIOUS: OnboardingMachineStateType.showProgress,
+      SET_PIN_CODE: {actions: assign({pinCode: (_, event) => event.data})},
+    },
+  },
+  verifyPinCode: {
+    on: {
+      NEXT: OnboardingMachineStateType.enableBiometrics,
+      PREVIOUS: OnboardingMachineStateType.enterPinCode,
+    },
+  },
+  enableBiometrics: {
+    on: {
+      NEXT: OnboardingMachineStateType.acceptTermsAndPrivacy,
+      PREVIOUS: OnboardingMachineStateType.verifyPinCode,
```

Comment:
hmmm i don;t think this should go to verifyPinCode, but to enterPinCode

Reply:
ditto